File: app/src/main/java/sixue/naivereader/BookshelfActivity.java
package sixue.naivereader;

import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.DialogInterface;
import android.content.Intent;
import android.content.IntentFilter;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.os.Bundle;
import android.support.design.widget.FloatingActionButton;
import android.support.v4.widget.SwipeRefreshLayout;
import android.support.v7.app.ActionBar;
import android.support.v7.app.AlertDialog;
import android.support.v7.app.AppCompatActivity;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.widget.AdapterView;
import android.widget.ArrayAdapter;
import android.widget.BaseAdapter;
import android.widget.CheckBox;
import android.widget.EditText;
import android.widget.GridView;
import android.widget.ImageView;
import android.widget.Spinner;
import android.widget.SpinnerAdapter;
import android.widget.TextView;

import java.util.ArrayList;
import java.util.List;

import sixue.naivereader.data.Book;
import sixue.naivereader.data.BookKind;
import sixue.naivereader.data.Source;

public class BookshelfActivity extends AppCompatActivity {

    private MyAdapter myAdapter;
    private boolean isEditMode;
    private List<Book> editList;
    private ActionBar actionBar;
    private BroadcastReceiver receiver;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_bookshelf);

        Utils.verifyPermissions(this);

        BookLoader.getInstance().reload(this);

        isEditMode = false;
        editList = new ArrayList<>();
        actionBar = getSupportActionBar();

        GridView gv = findViewById(R.id.gridview_books);
        final FloatingActionButton fab = findViewById(R.id.fab_add);
        final SwipeRefreshLayout srl = findViewById(R.id.srl);

        myAdapter = new MyAdapter();
        gv.setAdapter(myAdapter);
        gv.setOnItemClickListener(new AdapterView.OnItemClickListener() {
            @Override
            public void onItemClick(AdapterView<?> adapterView, View view, int i, long l) {
                if (!isEditMode) {
                    BookLoader.getInstance().bookBubble(i);

                    Intent intent = new Intent(BookshelfActivity.this, ReadActivity.class);
                    intent.setFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP);
                    startActivity(intent);
                } else {
                    View selectIcon = view.findViewById(R.id.select_icon);
                    boolean checked = !selectIcon.isSelected();
                    checkItem(selectIcon, i, checked);
                }
            }
        });
        gv.setOnItemLongClickListener(new AdapterView.OnItemLongClickListener() {
            @Override
            public boolean onItemLongClick(AdapterView<?> adapterView, View view, int i, long l) {
                if (!isEditMode) {
                    setEditMode(true);
                    View selectIcon = view.findViewById(R.id.select_icon);
                    checkItem(selectIcon, i, true);
                    return true;
                }
                return false;
            }
        });
        fab.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                if (!isEditMode) {
                    Intent intent = new Intent(BookshelfActivity.this, AddActivity.class);
                    startActivity(intent);
                }
            }
        });
        srl.setOnRefreshListener(new SwipeRefreshLayout.OnRefreshListener() {
            @Override
            public void onRefresh() {
                runOnUiThread(new Runnable() {
                    @Override
                    public void run() {
                        refreshAllBooks();
                        srl.setRefreshing(false);
                    }
                });
            }
        });

        receiver = new BroadcastReceiver() {

            @Override
            public void onReceive(Context context, Intent intent) {
                myAdapter.notifyDataSetChanged();
            }
        };
        IntentFilter filter = new IntentFilter();
        filter.addAction(Utils.ACTION_DOWNLOAD_CONTENT_FINISH);
        filter.addAction(Utils.ACTION_DOWNLOAD_COVER_FINISH);
        registerReceiver(receiver, filter);

        refreshAllBooks();
    }

    private void refreshAllBooks() {
        for (int i = 0; i < BookLoader.getInstance().getBookNum(); i++) {
            Book book = BookLoader.getInstance().getBook(i);
            if (book.getKind() == BookKind.Online) {
                SmartDownloader downloader = new SmartDownloader(this, book);
                if (downloader.reloadContent()) {
                    Intent intent = new Intent(Utils.ACTION_DOWNLOAD_CONTENT_FINISH);
                    intent.putExtra(Utils.INTENT_PARA_BOOK_ID, book.getId());
                    sendBroadcast(intent);
                }

                if (!book.isEnd()) {
                    downloader.startDownloadContent();
                }

                if (!downloader.coverIsDownloaded()) {
                    downloader.startDownloadContent();
                }
            }
        }
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        unregisterReceiver(receiver);
    }

    @Override
    public void onStart() {
        super.onStart();
        myAdapter.notifyDataSetChanged();
    }

    private void checkItem(View selectIcon, int i, boolean checked) {
        selectIcon.setSelected(checked);
        if (checked) {
            editList.add(BookLoader.getInstance().getBook(i));
        } else {
            editList.remove(BookLoader.getInstance().getBook(i));
        }
        invalidateOptionsMenu();
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.bookshelf, menu);
        return super.onCreateOptionsMenu(menu);
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem menuItem) {
        switch (menuItem.getItemId()) {
            case R.id.menu_delete:
                if (isEditMode) {
                    BookLoader.getInstance().deleteBooks(editList);
                    editList.clear();
                    setEditMode(false);
                    return true;
                }
                break;
            case R.id.menu_edit:
                if (editList.size() == 1) {
                    editItem(editList.get(0));
                    editList.clear();
                    setEditMode(false);
                    return true;
                }
                break;
            case R.id.menu_settings:
                Intent intent = new Intent(this, NetProviderManagerActivity.class);
                startActivity(intent);
                return true;
            default:
                break;
        }
        return super.onOptionsItemSelected(menuItem);
    }

    private void editItem(final Book book) {
        if (book.getKind() == BookKind.LocalText) {
            View v = getLayoutInflater().inflate(R.layout.edit_dialog_local, null);

            final EditText title = v.findViewById(R.id.title);
            title.setText(book.getTitle());
            title.clearFocus();

            final EditText author = v.findViewById(R.id.author);
            author.setText(book.getAuthor());
            author.clearFocus();

            EditText localPath = v.findViewById(R.id.local_path);
            localPath.setHint(book.getLocalPath());
            localPath.clearFocus();

            new AlertDialog.Builder(this)
                    .setTitle("Local book")
                    .setView(v)
                    .setPositiveButton(android.R.string.ok, new DialogInterface.OnClickListener() {
                        @Override
                        public void onClick(DialogInterface dialogInterface, int i) {
                            book.setTitle(title.getText().toString());
                            book.setAuthor(author.getText().toString());
                            BookLoader.getInstance().save();
                        }
                    })
                    .setNegativeButton(android.R.string.cancel, null)
                    .show();
        } else if (book.getKind() == BookKind.Online) {
            View v = getLayoutInflater().inflate(R.layout.edit_dialog_net, null);

            final EditText title = v.findViewById(R.id.title);
            title.setText(book.getTitle());
            title.clearFocus();

            final EditText author = v.findViewById(R.id.author);
            author.setText(book.getAuthor());
            author.clearFocus();

            Spinner sources = v.findViewById(R.id.sources);
            String[] sourceNames = Utils.convert(book.getSources(), new Utils.Func<Source>() {
                @Override
                public String exec(Source source) {
                    return source.getId();
                }
            });
            SpinnerAdapter adapter = new ArrayAdapter<>(this, android.R.layout.simple_spinner_item, sourceNames);
            sources.setAdapter(adapter);
            sources.clearFocus();

            final CheckBox end = v.findViewById(R.id.end);
            end.setChecked(book.isEnd());
            end.clearFocus();

            new AlertDialog.Builder(this)
                    .setTitle("Net book")
                    .setView(v)
                    .setPositiveButton(android.R.string.ok, new DialogInterface.OnClickListener() {
                        @Override
                        public void onClick(DialogInterface dialogInterface, int i) {
                            book.setTitle(title.getText().toString());
                            book.setAuthor(author.getText().toString());
                            book.setEnd(end.isChecked());
                            BookLoader.getInstance().save();
                        }
                    })
                    .setNegativeButton(android.R.string.cancel, null)
                    .show();
        }
    }

    @Override
    public boolean onPrepareOptionsMenu(Menu menu) {
        MenuItem edit = menu.findItem(R.id.menu_edit);
        MenuItem delete = menu.findItem(R.id.menu_delete);
        MenuItem settings = menu.findItem(R.id.menu_settings);

        edit.setVisible(isEditMode);
        delete.setVisible(isEditMode);
        settings.setVisible(!isEditMode);

        edit.setEnabled(editList.size() == 1);
        delete.setEnabled(editList.size() >= 1);

        return super.onPrepareOptionsMenu(menu);
    }

    private void setEditMode(boolean editMode) {
        isEditMode = editMode;
        myAdapter.notifyDataSetChanged();
        actionBar.setDisplayShowTitleEnabled(!editMode);
        invalidateOptionsMenu();
    }

    @Override
    public void onBackPressed() {
        if (isEditMode) {
            editList.clear();
            setEditMode(false);
            return;
        }

        super.onBackPressed();
    }

    private class MyAdapter extends BaseAdapter {

        MyAdapter() {

        }

        @Override
        public int getCount() {
            return BookLoader.getInstance().getBookNum();
        }

        @Override
        public Object getItem(int i) {
            return null;
        }

        @Override
        public long getItemId(int i) {
            return 0;
        }

        @Override
        public View getView(int i, View view, ViewGroup viewGroup) {
            if (view == null) {
                view = LayoutInflater.from(BookshelfActivity.this).inflate(R.layout.gridviewitem_book, viewGroup, false);
            }
            Book book = BookLoader.getInstance().getBook(i);

            TextView title = view.findViewById(R.id.title);
            title.setText(book.getTitle());

            TextView progress = view.findViewById(R.id.progress);
            if (book.getKind() == BookKind.LocalText) {
                int cp = book.getCurrentPosition();
                int wc = book.getWordCount();
                if (wc <= 0) {
                    progress.setText(R.string.read_progress_local_unread);
                } else {
                    progress.setText(getString(R.string.read_progress_local, (float) cp / (float) wc * 100f));
                }
            } else if (book.getKind() == BookKind.Online) {
                int size = book.getChapterList().size();
                int cp = book.getCurrentChapterIndex();
                if (size <= 0) {
                    progress.setText(R.string.read_progress_net_predownload);
                } else if (cp + 1 == size) {
                    progress.setText(R.string.read_progress_net_allread);
                } else {
                    progress.setText(getString(R.string.read_progress_net, size - cp - 1));
                }
            } else if (book.getKind() == BookKind.Packet) {
                int cp = book.getCurrentChapterIndex();
                progress.setText(getString(R.string.read_progress_net,  cp));
            }

            View selectIcon = view.findViewById(R.id.select_icon);
            if (!isEditMode) {
                selectIcon.setVisibility(View.INVISIBLE);
            } else {
                selectIcon.setVisibility(View.VISIBLE);
            }
            selectIcon.setSelected(editList.contains(book));

            ImageView cover = view.findViewById(R.id.cover);
            SmartDownloader downloader = new SmartDownloader(BookshelfActivity.this, book);
            if (downloader.coverIsDownloaded()) {
                Bitmap bm = BitmapFactory.decodeFile(book.getCoverSavePath());
                cover.setImageBitmap(bm);
            } else {
                Bitmap bm = Utils.getAutoCover(BookshelfActivity.this, book.getTitle());
                cover.setImageBitmap(bm);
            }
            return view;
        }
    }
}

File: app/src/main/java/sixue/naivereader/BookshelfActivity.kt
package sixue.naivereader

import android.content.*
import android.net.Uri
import android.os.Bundle
import android.view.*
import android.widget.*
import android.widget.AdapterView.OnItemClickListener
import android.widget.AdapterView.OnItemLongClickListener
import androidx.activity.OnBackPressedCallback
import androidx.activity.OnBackPressedDispatcher
import androidx.activity.result.ActivityResultLauncher
import androidx.activity.result.contract.ActivityResultContracts
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.app.AppCompatActivity
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout
import com.google.android.material.floatingactionbutton.FloatingActionButton
import sixue.naivereader.data.Book
import sixue.naivereader.data.BookKind
import sixue.naivereader.helper.ArchiveLoader
import java.util.*

class BookshelfActivity : AppCompatActivity() {
    private lateinit var myAdapter: MyAdapter
    private lateinit var editList: MutableList<Book>
    private lateinit var receiver: BroadcastReceiver
    private lateinit var getTextDocument: ActivityResultLauncher<Array<String>>
    private var updateBookCallback: ((Uri?) -> Unit)? = null
    private var isEditMode = false
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_bookshelf)
        Utils.verifyPermissions(this)
        BookLoader.reload(this)
        isEditMode = false
        editList = ArrayList()
        getTextDocument = registerForActivityResult(ActivityResultContracts.OpenDocument()) { uri: Uri? ->
            updateBookCallback?.invoke(uri)
        }
        val gv = findViewById<GridView>(R.id.gridview_books)
        val fab = findViewById<FloatingActionButton>(R.id.fab_add)
        val srl = findViewById<SwipeRefreshLayout>(R.id.srl)
        myAdapter = MyAdapter()
        gv.adapter = myAdapter
        gv.onItemClickListener = OnItemClickListener { _, view, i, _ ->
            if (!isEditMode) {
                BookLoader.bookBubble(i)
                val intent = Intent(this@BookshelfActivity, ReadActivity::class.java)
                intent.flags = Intent.FLAG_ACTIVITY_CLEAR_TOP
                startActivity(intent)
            } else {
                val selectIcon = view.findViewById<View>(R.id.select_icon)
                val checked = !selectIcon.isSelected
                checkItem(selectIcon, i, checked)
            }
        }
        gv.onItemLongClickListener = OnItemLongClickListener { _, view, i, _ ->
            if (!isEditMode) {
                setEditMode(true)
                val selectIcon = view.findViewById<View>(R.id.select_icon)
                checkItem(selectIcon, i, true)
                return@OnItemLongClickListener true
            }
            false
        }
        fab.setOnClickListener {
            if (!isEditMode) {
                val intent = Intent(this@BookshelfActivity, AddActivity::class.java)
                startActivity(intent)
            }
        }
        srl.setOnRefreshListener {
            runOnUiThread {
                refreshAllBooks()
                srl.isRefreshing = false
            }
        }
        receiver = object : BroadcastReceiver() {
            override fun onReceive(context: Context, intent: Intent) {
                myAdapter.notifyDataSetChanged()
            }
        }
        val filter = IntentFilter()
        filter.addAction(Utils.ACTION_DOWNLOAD_CONTENT_FINISH)
        filter.addAction(Utils.ACTION_DOWNLOAD_COVER_FINISH)
        registerReceiver(receiver, filter)
        refreshAllBooks()
    }

    private fun refreshAllBooks() {
        for (i in 0 until BookLoader.bookNum) {
            val book = BookLoader.getBook(i)
            if (book?.isRefreshable() == true) {
                if (book.buildHelper().reloadContent(this)) {
                    val intent = Intent(Utils.ACTION_DOWNLOAD_CONTENT_FINISH)
                    intent.putExtra(Utils.INTENT_PARA_BOOK_ID, book.id)
                    sendBroadcast(intent)
                }
                val downloader = SmartDownloader(this, book)
                if (!book.end) {
                    downloader.startDownloadContent()
                }
//                if (book.buildHelper().loadCoverBitmap(this) == null) {
//                    downloader.startDownloadContent()
//                }
            }
        }
    }

    public override fun onDestroy() {
        super.onDestroy()
        unregisterReceiver(receiver)
    }

    public override fun onStart() {
        super.onStart()
        myAdapter.notifyDataSetChanged()
        val callback = object : OnBackPressedCallback(true) {
            override fun handleOnBackPressed() {
                if (isEditMode) {
                    editList.clear()
                    setEditMode(false)
                }
            }
        }
        onBackPressedDispatcher.addCallback(this, callback)
    }

    private fun checkItem(selectIcon: View, i: Int, checked: Boolean) {
        selectIcon.isSelected = checked
        if (checked) {
            BookLoader.getBook(i)?.let { editList.add(it) }
        } else {
            editList.remove(BookLoader.getBook(i))
        }
        invalidateOptionsMenu()
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.bookshelf, menu)
        return super.onCreateOptionsMenu(menu)
    }

    override fun onOptionsItemSelected(menuItem: MenuItem): Boolean {
        when (menuItem.itemId) {
            R.id.menu_delete -> if (isEditMode) {
                BookLoader.deleteBooks(editList)
                editList.clear()
                setEditMode(false)
                return true
            }
            R.id.menu_edit -> if (editList.size == 1) {
                editItem(editList[0])
                editList.clear()
                setEditMode(false)
                return true
            }
            R.id.menu_settings -> {
                val intent = Intent(this, NetProviderManagerActivity::class.java)
                startActivity(intent)
                return true
            }
            else -> {}
        }
        return super.onOptionsItemSelected(menuItem)
    }

    private fun editItem(book: Book) {
        if (book.kind == BookKind.LocalText) {
            val v = View.inflate(this, R.layout.edit_dialog_local, null)
            val title = v.findViewById<EditText>(R.id.title)
            title.setText(book.title)
            title.clearFocus()
            val author = v.findViewById<EditText>(R.id.author)
            author.setText(book.author)
            author.clearFocus()
            val localPath = v.findViewById<EditText>(R.id.local_path)
            localPath.hint = book.localPath
            localPath.clearFocus()
            val browser = v.findViewById<Button>(R.id.button_browser)
            browser.setOnClickListener {
                getTextDocument.launch(arrayOf("text/plain"))
                localPath.setHint(R.string.HintTextEditing)
            }
            val dialog = AlertDialog.Builder(this)
                    .setTitle("Local book")
                    .setView(v)
                    .setPositiveButton(android.R.string.ok) { _, _ ->
                        book.title = title.text.toString()
                        book.author = author.text.toString()
                        BookLoader.save()
                    }
                    .setNegativeButton(android.R.string.cancel, null)
                    .show()
            updateBookCallback = { uri: Uri? ->
                if (uri != null) {
                    val takeFlags =
                        Intent.FLAG_GRANT_READ_URI_PERMISSION or Intent.FLAG_GRANT_WRITE_URI_PERMISSION
                    contentResolver.takePersistableUriPermission(uri, takeFlags)
                    book.localPath = uri.toString()
                    BookLoader.save()
                    localPath.hint = uri.toString()
                } else {
                    localPath.hint = book.localPath
                }
                dialog.dismiss()
            }
        } else if (book.kind == BookKind.Online) {
            val v = View.inflate(this, R.layout.edit_dialog_net, null)
            val title = v.findViewById<EditText>(R.id.title)
            title.setText(book.title)
            title.clearFocus()
            val author = v.findViewById<EditText>(R.id.author)
            author.setText(book.author)
            author.clearFocus()
            val sources = v.findViewById<Spinner>(R.id.sources)
            val sourceNames = book.sources.map { it.id }
            val adapter: SpinnerAdapter = ArrayAdapter(this, android.R.layout.simple_spinner_item, sourceNames)
            sources.adapter = adapter
            sources.clearFocus()
            val end = v.findViewById<CheckBox>(R.id.end)
            end.isChecked = book.end
            end.clearFocus()
            val archive = v.findViewById<Button>(R.id.archive)
            val dialog = AlertDialog.Builder(this)
                    .setTitle("Net book")
                    .setView(v)
                    .setPositiveButton(android.R.string.ok) { _, _ ->
                        book.title = title.text.toString()
                        book.author = author.text.toString()
                        book.end = end.isChecked
                        BookLoader.save()
                    }
                    .setNegativeButton(android.R.string.cancel, null)
                    .show()
            archive.setOnClickListener {
                val newBook = ArchiveLoader.createArchive(book, this)
                BookLoader.replace(book, newBook)
                myAdapter.notifyDataSetChanged()
                dialog.dismiss()
            }
        }
    }

    override fun onPrepareOptionsMenu(menu: Menu): Boolean {
        val edit = menu.findItem(R.id.menu_edit)
        val delete = menu.findItem(R.id.menu_delete)
        val settings = menu.findItem(R.id.menu_settings)
        edit.isVisible = isEditMode
        delete.isVisible = isEditMode
        settings.isVisible = !isEditMode
        edit.isEnabled = editList.size == 1
        delete.isEnabled = editList.size >= 1
        return super.onPrepareOptionsMenu(menu)
    }

    private fun setEditMode(editMode: Boolean) {
        isEditMode = editMode
        myAdapter.notifyDataSetChanged()
        supportActionBar?.setDisplayShowTitleEnabled(!editMode)
        invalidateOptionsMenu()
    }

    private inner class MyAdapter : BaseAdapter() {
        override fun getCount(): Int {
            return BookLoader.bookNum
        }

        override fun getItem(i: Int): Any {
            return ""
        }

        override fun getItemId(i: Int): Long {
            return 0
        }

        override fun getView(i: Int, convertView: View?, viewGroup: ViewGroup): View {
            val view : View?
            if (convertView == null) {
                view = LayoutInflater.from(this@BookshelfActivity).inflate(R.layout.gridviewitem_book, viewGroup, false)
                val title = view.findViewById<TextView>(R.id.title)
                val progress = view.findViewById<TextView>(R.id.progress)
                val selectIcon = view.findViewById<View>(R.id.select_icon)
                val cover = view.findViewById<ImageView>(R.id.cover)
                view.tag = ViewHolder(title, progress, selectIcon, cover)
            } else {
                view = convertView
            }
            val viewHolder : ViewHolder = (view!!.tag as ViewHolder)

            val book = BookLoader.getBook(i)
            viewHolder.title.text = book?.title
            viewHolder.progress.text = book?.buildHelper()?.progressText(this@BookshelfActivity)
            if (!isEditMode) {
                viewHolder.selectIcon.visibility = View.INVISIBLE
            } else {
                viewHolder.selectIcon.visibility = View.VISIBLE
            }
            viewHolder.selectIcon.isSelected = editList.contains(book)
            viewHolder.cover.setImageBitmap(book?.buildHelper()?.loadCoverBitmap(this@BookshelfActivity))
            return view
        }

        inner class ViewHolder(val title: TextView, val progress: TextView, val selectIcon: View, val cover: ImageView)
    }
}